### PR TITLE
[NFC] IRGen: Extracted TI::withMetadataCollector.

### PR DESCRIPTION
**Pull request by nate-chandler**
Deduplicate some code.

### CODE REVIEW

File: lib/IRGen/Outlining.cpp [hunk 1]
```diff
@@ -67,18 +73,24 @@ void OutliningMetadataCollector::collectTypeMetadataForLayout(SILType type) {
   Values.insert({key, metadata});
 }
 
-void OutliningMetadataCollector::collectFormalTypeMetadata(CanType type) {
-  // If the type has no archetypes, we can emit it from scratch in the callee.
-  assert(type->hasArchetype());
+void OutliningMetadataCollector::collectTypeMetadataForEnvironment(SILType ty) {
+  auto bgt = dyn_cast<BoundGenericType>(ty.getASTType());
```
Comment (slavapestov) [inline lib/IRGen/Outlining.cpp:77]: This is insufficient:

- The BoundGenericType `Outer<Int>.Inner<String>` has a parent type `Outer<Int>` you must visit:
```
struct Outer<T> { struct Inner<U> { } }
```

- A NominalType might have a parent type:
```
struct Outer<T> { struct Inner { } }
```

Instead, you need to walk the replacement types in the context substitution map:

```
if (type->isSpecialized()) {
  auto subMap = type->getContextSubstitutionMap(module, type->getAnyNominal());
  for (auto replacementType : subMap.getReplacementTypes()) { ... }
}
```

Also, what about witness tables? The generic type might depend on a conformance:

```
struct Foo<T: P> {
  var x: T.A

  deinit {
    ... doing stuff with x requires T: P
  }
}
```

**Review by slavapestov — CHANGES_REQUESTED** [review]

File: lib/IRGen/Outlining.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/IRGen/Outlining.cpp:77]: Bonus: type in constrained extension depends on the constrained extension's conformance:

```
struct G<T> {}

extension G where T: Sequence {
  struct Inner: ~Copyable {
    var x: T.Element

    deinit { ... }
  }
}
```

Comment (nate-chandler) [reply lib/IRGen/Outlining.cpp:77]: Yes, I am aware of these limitations.

File: test/IRGen/moveonly_enum_deinits.swift [hunk 2]
```diff
@@ -1,9 +1,49 @@
...
+
+// Check that the outlined retain function has a parameter for the type's
+// generic parameter and that it's passed along to the deinit.
+// CHECK-LABEL: define{{.*}} ptr @"$s21moveonly_enum_deinits17UnsafeManualArrayVyxGlWOs"(
+// CHECK-SAME:      ptr %0, 
+// CHECK-SAME:      ptr %T)
+// CHECK-SAME:  {
+// CHECK:         call swiftcc void @"$s21moveonly_enum_deinits17UnsafeManualArrayVfD"(
+// CHECK-SAME:        ptr %T)
+// CHECK:       }
```
Comment (jckarter) [inline test/IRGen/moveonly_enum_deinits.swift:33]: There's no point in an outlined release that just calls a `deinit` that's already a separate function. How hard would it be to not outline these value witness operations at all for a noncopyable type that directly has a `deinit`?

**Comment by nate-chandler** [comment]
@swift-ci please smoke test

File: test/IRGen/moveonly_enum_deinits.swift (hunk 2 above)
Comment (nate-chandler) [reply test/IRGen/moveonly_enum_deinits.swift]: That's a nice peephole.  Implemented here: https://github.com/apple/swift/pull/71787 .  There's a similar one to do for the destroy.

Unfortunately, it's not always applicable.  Generic move-only structs _without_ deinits that contain generic move-only structs _with_ deinits should still get outlined value functions for release and destroy and those outlined value functions need the metadata to pass to the deinits of the contained move-only structs with deinits.  That's implemented here: https://github.com/apple/swift/pull/71799

**Pull request merged**